src/content/faq/tools/jtag.md

```diff
@@ -0,0 +1,369 @@
+
+
+## Programmers 
+
+The standard programmer is the [ST-LINK/V2](http://www.st.com/en/development-tools/st-link-v2.html). It connects to your computer using USB and to the board using JTAG or SWD.
+
+![ST-LINK/V2](/assets/images/jtag-01stlink.jpg)
+
+There are also "ST-LINK/V2 Mini" devices. These also connect by USB but only use the SWD interface. These inexpensive clone devices are available on Amazon and eBay for US$12 or even less.
+
+![ST-LINK/V2 Mini](/assets/images/jtag-07mini.jpg)
+
+Finally, there's the [Particle Programmer Shield](https://github.com/spark/shields/tree/master/photon-shields/programmer-shield), primarily designed for the Photon.  
+
+![Particle Programmer Shield](/assets/images/jtag-08shield.jpg)
+
+
+## Connecting to the Photon/Electron/P1
+
+Since the JTAG/SWD interface uses the same pins as your project may use, it will probably be necessary to stop your code from running so it won't reinitialize the pins needed for JTAG/SWD. The easiest way is to enable DFU mode (blinking yellow). This works even if you have a P1 without a USB port. 
+
+Even if you don't use the pins, you will likely have problems using SWD in normal operating mode (breathing cyan).
+
+### JTAG
+
+Pins:
+
+- D7: TMS
+- D6: TCK
+- D5: TDI
+- D4: TDO
+- D3: TRST
+- GND
+
+### SWD
+
+Pins:
+
+- D7: SWDIO
+- D6: SWCLK
+- GND
+
+### Programmer shield
+
+Since you plug the Photon into the programmer shield, there's nothing to connect.
+
+
+## Using ST/LINK for Windows
+
+By far the easiest way to program the flash is to use the ST/LINK application for Windows. It's a free download from [ST](http://www.st.com/en/development-tools/st-link-v2.html), and works with both the real ST/LINK device and many clone devices. 
+
+For the Mac, Linux, and for the Particle Programmer shield you'll need to use OpenOCD, in the next section.
+
+
+### Programming the boot loader
+
+One of the common uses for a programmer is fixing the situation where the boot loader gets corrupted. One symptom of a corrupted boot loader is the Photon/Electron only has a dim D7 LED and doesn't respond to any other buttons. And the main LED does not light.
+
+Connect the ST/LINK JTAG/SWD programmer as listed above and run the **STM32 ST-LINK Utility** Windows application.
+
+Select **Connect** from the **Target** menu or use the toolbar icon. It should look something like this:
+
+![ST-LINK Connect](/assets/images/jtag-02connect.png)
+
+The bootloader is different for each type of device and changes infrequently. Download the appropriate one for your device:
+
+- [Photon](https://github.com/rickkas7/particle_jtag/blob/master/files/bootloader-photon.bin?raw=true) (0.4.9)
+- [P1](https://github.com/rickkas7/particle_jtag/blob/master/files/bootloader-p1.bin?raw=true) (0.4.9)
+- [Electron](https://github.com/rickkas7/particle_jtag/blob/master/files/bootloader-electron.bin?raw=true) (0.4.8)
```

Comment:
I'd like to make this more future-proof.  Those Photon/P1 bootloaders kind of feel like they might already be outdated if they are not v7.